src/blockedtuple.jl

```diff
@@ -0,0 +1,76 @@
+# This file defines BlockedTuple, a Tuple of heterogeneous Tuple with a BlockArrays.jl
+# like interface
+
+using BlockArrays: Block, BlockArrays, BlockIndexRange, BlockRange, blockedrange
+
+struct BlockedTuple{Divs,Flat}
+  flat::Flat
+
+  function BlockedTuple{Divs}(flat::Tuple) where {Divs}
+    length(flat) != sum(Divs) && throw(DimensionMismatch("Invalid total length"))
+    return new{Divs,typeof(flat)}(flat)
+  end
+end
+
+# TensorAlgebra Interface
+BlockedTuple(tt::Vararg{Tuple}) = BlockedTuple{length.(tt)}(flatten_tuples(tt))
+BlockedTuple(bt::BlockedTuple) = bt
+flatten_tuples(bt::BlockedTuple) = Tuple(bt)
+
+# Base interface
+Base.Tuple(bt::BlockedTuple) = bt.flat
+
+Base.axes(bt::BlockedTuple) = (blockedrange([blocklengths(bt)...]),)
+
+Base.broadcastable(bt::BlockedTuple) = bt
+struct BlockedTupleBroadcastStyle{Divs} <: Broadcast.BroadcastStyle end
+function Base.BroadcastStyle(::Type{<:BlockedTuple{Divs}}) where {Divs}
+  return BlockedTupleBroadcastStyle{Divs}()
+end
```

Comment:
```suggestion
function Base.BroadcastStyle(::Type{<:BlockedTuple{BlockLengths}}) where {BlockLengths}
  return BlockedTupleBroadcastStyle{BlockLengths}()
end
```
to keep naming conventions consistent.

Reply:
Also, we can make this more generic by writing it as:
```julia
function Base.BroadcastStyle(type::Type{<:AbstractBlockTuple})
  return BlockedTupleBroadcastStyle{blocklengths(type)}()
end
```
and then define `blocklengths(::Type{<:AbstractBlockTuple})` for the various `AbstractBlockTuple` subtypes.

Reply:
So there is a difficulty here. Ultimately, I need a constructor to return the correct object at the end. This is done in
```julia
function Base.copy(
  bc::Broadcast.Broadcasted{BlockedTupleBroadcastStyle{BlockLengths}}
) where {BlockLengths}
  return BlockedTuple{BlockLengths}(bc.f.((Tuple.(bc.args))...))
end
```
If I share the broadcast style with `AbstractBlockTuple`, then I need a way to dispatch in `copy`

I found a solution by adding the wrapper type as a style parameter:
```julia
# Broadcast interface
Base.broadcastable(bt::AbstractBlockTuple) = bt
struct AbstractBlockTupleBroadcastStyle{BlockLengths,BT} <: Broadcast.BroadcastStyle end
function Base.BroadcastStyle(type::Type{<:AbstractBlockTuple})
  return AbstractBlockTupleBroadcastStyle{blocklengths(type),type.name.wrapper}()
end

# BroadcastStyle is not called for two identical styles
function Base.BroadcastStyle(::AbstractBlockTupleBroadcastStyle, ::AbstractBlockTupleBroadcastStyle)
  throw(DimensionMismatch("Incompatible blocks"))
end
function Base.copy(
  bc::Broadcast.Broadcasted{AbstractBlockTupleBroadcastStyle{BlockLengths,BT}}
) where {BlockLengths,BT}
  return BT{BlockLengths}(bc.f.((Tuple.(bc.args))...))
end
```
but I am not super happy with it.

Reply:
That seems like a reasonable solution.

It is a tricky thing in Julia because normally in Julia one would use `similar` to determine the output storage type, but that mostly makes sense for mutable storage types.

For a nicer API for getting the "base type" (i.e. alternative to `type.name.wrapper`), you can use [`TypeParameterAccessors.unspecify_type_parameters`](https://github.com/ITensor/TypeParameterAccessors.jl).

Reply:
We could use the same code style to make `map` generic to `AbstractBlockTuple`.

Reply:
I was considering subtyping `AbstractBlockTupleBroadcastStyle` indeed, but as you mentioned it requires custom code for each subtype. I will use a type parameter accessed with `TypeParameterAccessors`.